src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs

```diff
@@ -25,14 +25,67 @@ internal enum TypeParameterConstraintKind
+        internal static readonly TypeParameterConstraintClause Empty = new TypeParameterConstraintClause(
+            TypeParameterConstraintKind.None,
+            ImmutableArray<TypeSymbolWithAnnotations>.Empty,
+            syntax: default,
+            otherPartialDeclarations: ImmutableArray<TypeParameterConstraintClause>.Empty);
+
+        internal static TypeParameterConstraintClause Create(
+            TypeParameterConstraintKind constraints,
+            ImmutableArray<TypeSymbolWithAnnotations> constraintTypes,
+            ImmutableArray<TypeConstraintSyntax> syntax = default)
         {
             Debug.Assert(!constraintTypes.IsDefault);
+            if (constraints == TypeParameterConstraintKind.None && constraintTypes.IsEmpty)
+            {
+                Debug.Assert(syntax.IsDefault);
+                return Empty;
+            }
+            return new TypeParameterConstraintClause(constraints, constraintTypes, syntax, otherPartialDeclarations: ImmutableArray<TypeParameterConstraintClause>.Empty);
+        }
+
+        private TypeParameterConstraintClause(
+            TypeParameterConstraintKind constraints,
+            ImmutableArray<TypeSymbolWithAnnotations> constraintTypes,
+            ImmutableArray<TypeConstraintSyntax> syntax,
+            ImmutableArray<TypeParameterConstraintClause> otherPartialDeclarations)
+        {
             this.Constraints = constraints;
             this.ConstraintTypes = constraintTypes;
+            this.Syntax = syntax;
+            this.OtherPartialDeclarations = otherPartialDeclarations;
         }
 
         public readonly TypeParameterConstraintKind Constraints;
         public readonly ImmutableArray<TypeSymbolWithAnnotations> ConstraintTypes;
+
+        /// <summary>
+        /// Syntax for the constraint types. Populated from early constraint checking step only.
+        /// </summary>
+        internal readonly ImmutableArray<TypeConstraintSyntax> Syntax;
+
+        /// <summary>
+        /// Collection of constraint clauses for other partial declarations of the same container.
+        /// Populated from early constraint checking step only.
+        /// </summary>
+        internal readonly ImmutableArray<TypeParameterConstraintClause> OtherPartialDeclarations;
+
+        internal bool IsEmpty => Constraints == TypeParameterConstraintKind.None && ConstraintTypes.IsEmpty;
+
+        internal bool IsEarly => !Syntax.IsDefault;
+
+        internal TypeParameterConstraintClause AddPartialDeclaration(TypeParameterConstraintClause other)
+        {
+            return new TypeParameterConstraintClause(Constraints, ConstraintTypes, Syntax, OtherPartialDeclarations.Add(other));
+        }
+    }
+
+    internal static class TypeParameterConstraintClauseExtensions
+    {
+        internal static bool IsEarly(this ImmutableArray<TypeParameterConstraintClause> constraintClauses)
+        {
+            return constraintClauses.Any(clause => clause.IsEarly == true);
```

Comment:
> == true [](start = 65, length = 8)

Do we need `== true` here?

Reply:
Not needed, thanks.

---
In reply to: [195548003](https://github.com/dotnet/roslyn/pull/27525#discussion_r195548003) [](ancestors = 195548003)